File: AutoMix Software/AutoMix AI/GeneticSortAlgorithm.cpp
// Copyright (C) 2016-2017 LesProjecteurs - All Rights Reserved
// Maxime STEVENOT, Guillaume HANNES, Jordan ERNULT, Louis CARLIER, Pierre GABON
// 
// This file is part of AutoMix Software.
// 
// Unauthorized copying of this file, via any medium is strictly prohibited.
// You should have received a copy of the License along with this program.

#include "stdafx.h"
#include "GeneticSortAlgorithm.h"

using namespace System;
using namespace AutoMixDataManagement;

namespace AutoMixAI
{
	GeneticSortAlgorithm::GeneticSortAlgorithm(TrackDistance^ distance) : GeneticSortAlgorithm(distance, 175, 175, 85) {}

	GeneticSortAlgorithm::GeneticSortAlgorithm(TrackDistance^ distance, unsigned int numberOfIteration, unsigned int populationSize, double mutationProbabilty)
		: SortAlgorithm(distance)
	{
		NUMBER_OF_ITERATION = numberOfIteration;
		POPULATION_SIZE = populationSize;
		MUTATION_PROBABILITY = mutationProbabilty;
	}

	TrackCollection^ GeneticSortAlgorithm::sort(System::ComponentModel::BackgroundWorker^ bw, TrackCollection^ trackCollection)
	{
		Population^ pop = createInitialPopulation(trackCollection);
		sortPopulation(pop, 0, pop->Count - 1);

		for (int k = 0; k < NUMBER_OF_ITERATION; k++)
		{
			if (bw->CancellationPending)
			{
				break;
			}

			System::Diagnostics::Debug::WriteLine("iteration {0}", k);
			System::Diagnostics::Debug::WriteLine("premier {0}", computeIndividualEvaluation(pop[0]));
			System::Diagnostics::Debug::WriteLine("dernier {0}", computeIndividualEvaluation(pop[POPULATION_SIZE - 1]));

			createChildAndPutThemIntoPopulation(pop);
			mutatePopulation(pop);
			sortPopulation(pop, 0, pop->Count - 1);
			pop = pop->GetRange(0, POPULATION_SIZE);
			bw->ReportProgress((int)1000 * k / NUMBER_OF_ITERATION);
		}
		return pop[0];
	}

	Population^ GeneticSortAlgorithm::createInitialPopulation(TrackCollection^ trackCollection)
	{
		Population^ pop = gcnew Population();
		System::Random^ rand = gcnew System::Random();
		pop->Add(trackCollection);
		for (int k = 1; k < POPULATION_SIZE; k++)
		{
			TrackCollection^ individual = gcnew TrackCollection();
			for (int i = 0; i < trackCollection->Count; i++)
			{
				individual->Add(trackCollection[i]);
			}
			for (int i = 0; i < individual->Count; i++)
			{
				Track^ track = individual[i];
				individual->Remove(track);
				individual->Insert(rand->Next(individual->Count), track);
			}
			pop->Add(individual);
		}
		return pop;
	}

	double GeneticSortAlgorithm::computeIndividualEvaluation(TrackCollection^ individual)
	{
		double result = 0;
		for (int k = 0; k < (individual->Count) - 1; k++)
		{
			result += Distance->compute(individual[k], individual[k + 1]);
			//result = result + computeTracksDistance(individual[k], individual[k + 1]);
		}
		return result;
	}

	void GeneticSortAlgorithm::sortPopulation(Population^ population, int begin, int end)
	{
		int i = begin;
		int j = end;
		TrackCollection^ temp;
		TrackCollection^ pivot = population[(i + j) / 2];

		while (i <= j)
		{
			while (computeIndividualEvaluation(population[i]) < computeIndividualEvaluation(pivot))
			{
				i++;
			}
			while (computeIndividualEvaluation(population[j]) > computeIndividualEvaluation(pivot))
			{
				j--;
			}
			if (i <= j)
			{
				temp = population[i];
				population[i] = population[j];
				population[j] = temp;
				i++;
				j--;
			}
		}
		if (begin < j)
		{
			sortPopulation(population, begin, j);
		}
		if (i < end)
		{
			sortPopulation(population, i, end);
		}
	}

	void GeneticSortAlgorithm::createChildAndPutThemIntoPopulation(Population^ pop)
	{
		int count = pop->Count;
		for (int k = 0; k < count / 2; k++)
		{
			TrackCollection^ children1 = createChildrenFromParents(pop[k], pop[k + 1]);
			TrackCollection^ children2 = createChildrenFromParents(pop[k + 1], pop[k]);
			pop->Add(children1);
			pop->Add(children2);
		}
	}

	TrackCollection^ GeneticSortAlgorithm::createChildrenFromParents(TrackCollection^ parent1, TrackCollection^ parent2)
	{
		TrackCollection^ children = gcnew TrackCollection();
		for (int k = 0; k < (parent1->Count) / 2; k++)
		{
			children->Add(parent1[k]);
		}

		for (int i = (parent1->Count) / 2; i < parent1->Count; i++)
		{
			if (children->Contains(parent2[i]))
			{
				for (int j = 0; j < parent1->Count; j++)
				{
					if ((!children->Contains(parent1[j])) && (!parent2->GetRange((parent1->Count) / 2, (parent1->Count) / 2)->Contains(parent1[j])))
					{
						children->Add(parent1[j]);
					}
				}
			}
			else
			{
				children->Add(parent2[i]);
			}
		}
		return children;
	}

	void GeneticSortAlgorithm::mutatePopulation(Population^ pop)
	{
		System::Random^ rand = gcnew System::Random();
		for (int k = 0; k < pop->Count; k++)
		{

			if (rand->Next(100) < MUTATION_PROBABILITY)
			{
				TrackCollection^ mutant = pop[k];
				Track^ temp;

				int index1 = rand->Next(mutant->Count);
				int index2 = rand->Next(mutant->Count);
				temp = mutant[index1];
				mutant[index1] = mutant[index2];
				mutant[index2] = temp;
				pop[k] = mutant;
			}
		}

	}
}

File: Automix AI/Sort Algorithms/GeneticSortAlgorithm.cs
// Copyright (C) 2016-2017 LesProjecteurs - All Rights Reserved
// Maxime STEVENOT, Guillaume HANNES, Jordan ERNULT, Louis CARLIER, Pierre GABON
// 
// This file is part of Automix Software.
// 
// Unauthorized copying of this file, via any medium is strictly prohibited.
// You should have received a copy of the License along with this program.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using Automix_AI.Distances;
using Automix_Data_Management.Model;

namespace Automix_AI.Sort_Algorithms
{
    using Population = List<TrackCollection>;

    public class GeneticSortAlgorithm : AbstractSortAlgorithm
    {
        private readonly int _numberOfIteration;
        private readonly int _populationSize;
        private readonly double _mutationProbability; // Probabilty of mutation out of 100

        public GeneticSortAlgorithm(ITracksDistance distance) : this(distance, 175, 175, 85) { }

        public GeneticSortAlgorithm(ITracksDistance distance, int numberOfIteration, int populationSize, double mutationProbabilty) : base(distance)
        {
            _numberOfIteration = numberOfIteration;
            _populationSize = populationSize;
            _mutationProbability = mutationProbabilty;
        }

        public override TrackCollection Sort(BackgroundWorker backgroundWorker, TrackCollection trackCollection, TrackCollection fixedTracks)
        {
            var population = CreateInitialPopulation(trackCollection);
            SortPopulation(population, 0, population.Count - 1);

            for (var k = 0; k < _numberOfIteration; k++)
            {
                if (backgroundWorker.CancellationPending)
                {
                    break;
                }

                CreateChildrenAndPutThemIntoPopulation(population);
                MutatePopulation(population);
                SortPopulation(population, 0, population.Count - 1);
                population = population.GetRange(0, _populationSize);
                backgroundWorker.ReportProgress(1000 * k / _numberOfIteration);
            }
            return population[0];
        }

        private void MutatePopulation(Population pop)
        {
            var  randomGenerator = new Random();
            for (var k = 0; k < pop.Count; k++)
            {
                if (!(randomGenerator.Next(100) < _mutationProbability))
                {
                    continue;
                }
                var mutant = pop[k];
                var index1 = randomGenerator.Next(mutant.Count);
                var index2 = randomGenerator.Next(mutant.Count);
                var tempTrack = mutant[index1];
                mutant[index1] = mutant[index2];
                mutant[index2] = tempTrack;
                pop[k] = mutant;
            }
        }

        private void CreateChildrenAndPutThemIntoPopulation(Population population)
        {
            for (var k = 0; k < population.Count / 2; k++)
            {
                population.Add(CreateChildrenFromParents(population[k], population[k + 1]));
                population.Add(CreateChildrenFromParents(population[k + 1], population[k]));
            }
        }

        private TrackCollection CreateChildrenFromParents(TrackCollection parent1, TrackCollection parent2)
        {
            var  children = new TrackCollection();
            for (var k = 0; k < (parent1.Count) / 2; k++)
            {
                children.Add(parent1[k]);
            }

            for (var i = (parent1.Count) / 2; i < parent1.Count; i++)
            {
                if (children.Contains(parent2[i]))
                {
                    foreach (var t in parent1)
                    {
                        if ((!children.Contains(t)) && (!parent2.GetRange((parent1.Count) / 2, (parent1.Count) / 2).Contains(t)))
                        {
                            children.Add(t);
                        }
                    }
                }
                else
                {
                    children.Add(parent2[i]);
                }
            }
            return children;
        }

        private void SortPopulation(Population population, int begin, int end)
        {
            var i = begin;
            var j = end;
            var pivot = population[(i + j) / 2];

            while (i <= j)
            {
                while (ComputeIndividualEvaluation(population[i]) < ComputeIndividualEvaluation(pivot))
                {
                    i++;
                }
                while (ComputeIndividualEvaluation(population[j]) > ComputeIndividualEvaluation(pivot))
                {
                    j--;
                }
                if (i <= j)
                {
                    var temp = population[i];
                    population[i] = population[j];
                    population[j] = temp;
                    i++;
                    j--;
                }
            }
            if (begin < j)
            {
                SortPopulation(population, begin, j);
            }
            if (i < end)
            {
                SortPopulation(population, i, end);
            }
        }

        private Population CreateInitialPopulation(TrackCollection trackCollection)
        {
            var population = new Population();
            var randomGenerator = new Random();
            population.Add(trackCollection);
            for (var k = 1; k < _populationSize; k++)
            {
                var individual = new TrackCollection();
                individual.AddRange(trackCollection);
                for (var i = 0; i < individual.Count; i++)
                {
                    var track = individual[i];
                    individual.Remove(track);
                    individual.Insert(randomGenerator.Next(individual.Count), track);
                }
                population.Add(individual);
            }
            return population;
        }

        private double ComputeIndividualEvaluation(TrackCollection individual)
        {
            double result = 0;
            for (var k = 0; k < (individual.Count) - 1; k++)
            {
                result += Distance.Compute(individual[k], individual[k + 1]);
            }
            return result;
        }
    }
}
